src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -1243,6 +1227,81 @@ lExit: ;
 }
 FCIMPLEND
 
+FCIMPL1(FC_BOOL_RET, RuntimeFieldHandle::IsFastPathSupported, ReflectFieldObject *pFieldUNSAFE)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    FieldDesc* pFieldDesc = refField->GetField();
+    return pFieldDesc->IsEnCNew() ? FALSE : TRUE;
+}
+FCIMPLEND
+
+FCIMPL1(INT32, RuntimeFieldHandle::GetInstanceFieldAddress, ReflectFieldObject *pFieldUNSAFE)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+        PRECONDITION(CheckPointer(pFieldUNSAFE));
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    FieldDesc* pFieldDesc = refField->GetField();
+
+    // IsFastPathSupported needs to checked before calling this method.
+    _ASSERTE(!pFieldDesc->IsEnCNew());
+
+    return pFieldDesc->GetOffset();
+}
+FCIMPLEND
+
+FCIMPL2(void*, RuntimeFieldHandle::GetStaticFieldAddress, ReflectFieldObject *pFieldUNSAFE, CLR_BOOL *isBoxed)
+{
+    CONTRACTL {
+        FCALL_CHECK;
+        PRECONDITION(CheckPointer(pFieldUNSAFE));
+    }
+    CONTRACTL_END;
+
+    REFLECTFIELDREF refField = (REFLECTFIELDREF)ObjectToOBJECTREF(pFieldUNSAFE);
+    if (refField == NULL)
+        FCThrowRes(kArgumentNullException, W("Arg_InvalidHandle"));
+
+    void *ret = NULL;
+    FieldDesc* pFieldDesc = refField->GetField();
+
+    // IsFastPathSupported needs to checked before calling this method.
+    _ASSERTE(!pFieldDesc->IsEnCNew());
+
+    GCPROTECT_BEGIN(refField);
+
+    if (refField->GetField()->IsThreadStatic()) {
+        ret = Thread::GetStaticFieldAddress(pFieldDesc);
```

Comment:
These addresses may be in GC allocated memory in some cases. They can move and need to be reported to the GC.